components/speedreader/speedreader_service.cc

```diff
@@ -6,33 +6,94 @@
 #include "brave/components/speedreader/speedreader_service.h"
 
 #include "base/feature_list.h"
+#include "base/metrics/histogram_macros.h"
 #include "brave/components/speedreader/features.h"
 #include "brave/components/speedreader/speedreader_pref_names.h"
+#include "brave/components/weekly_storage/weekly_storage.h"
 #include "components/prefs/pref_registry_simple.h"
 #include "components/prefs/pref_service.h"
 
 namespace speedreader {
 
+namespace {
+
+// Note: append-only enumeration! Never remove any existing values, as this enum
+// is used to bucket a UMA histogram, and removing values breaks that.
+constexpr std::array<uint64_t, 5> kSpeedReaderToggleBuckets{
+    0,   // 0
+    5,   // >0-5
+    10,  // 6-10
+    20,  // 11-20
+    30,  // 21-30
+         // 30+ => bucket 5
+};
+
+constexpr char kSpeedreaderToggleUMAHistogramName[] =
+    "Brave.SpeedReader.ToggleCount";
+
+constexpr char kSpeedreaderEnabledUMAHistogramName[] =
+    "Brave.SpeedReader.Enabled";
+
+void storeTogglesHistogram(uint64_t toggles) {
+  int bucket = 0;
+  for (const auto& bucketUpperBound : kSpeedReaderToggleBuckets) {
+    if (toggles > bucketUpperBound)
+      bucket = bucket + 1;
+  }
+
+  UMA_HISTOGRAM_EXACT_LINEAR(kSpeedreaderToggleUMAHistogramName, bucket, 5);
+}
+
+void storeSpeedReaderEnabledHistogram(PrefService* prefs_,
+                                      bool toggled,
+                                      bool enabledNow) {
+  if (toggled) {
+    WeeklyStorage weekly(prefs_, kSpeedreaderPrefToggleCount);
+    weekly.AddDelta(1);
+    storeTogglesHistogram(weekly.GetWeeklySum());
+  }
+
+  if (toggled && enabledNow) {
```

Comment:
instead of this, why not record the metric in `SpeedreaderTabHelper::UpdateActiveState`, where we are setting `active_` to true?
`Have you used SpeedReader?` sounds more like "Have you ever read distilled article" for me

Reply:
different semantics though:
- one tracks if user has ever or recently attempted to enable the feature
- the other one tracks if the feature did something

Explicitly want to know if users tried enabling it and just never came back because they didn't even see it working

Reply:
Sorry, now I'm confused. In the ticket I see
```
0) Never Used
1) Used but not recently (in reporting interval)
2) Used within the last week
```
Here, we record `2` when the feature was explicitly enabled by toggle. But if this function is never reached again (i.e. the browser stays open for a month without doing anything), we would send `2` every week, which is not "Used within the last week"

Next, `IsEnabled()` is hit for each navigation, so in real world we almost immediately will overwrite `2` with `1`, which is also something unexpected for me.

Reply:
Might be my misunderstanding of how P3A uses UMA metrics. I believe it:
- takes a list of tuples `(item, count)`, finds largest `item` with `count > 0`
- reports the `item`
- resets everything to 0

How would you do "used but not recently" then? Indeed if SpeedReader got enabled and stays enabled from one week to the next, it should report `2`. If it got enabled at some point, then disabled and stays disabled, it should report `1` for every week when it didn't get enabled (or stayed enabled) at all. Finally if it absolutely never got enabled, it should be `0`.

The intention is to track retention - a user that keeps the feature always enabled is _retained_. A user who tried the feature at some point and stopped using it is lost.

Reply:
> takes a list of tuples (item, count), finds largest item with count > 0
what do you mean by "largest" item?

P3A basically just scrubs the latest emitted value and saves it to the map "metric -> value", overwriting the previous value. It never resets anything on its own or chooses anything. Just holds and periodically transmits the last number.

Reply:
ok, changed the logic to track if the state is currently enabled _or_ it has been toggled this week at least once, implying that it was enabled at some point this week